File: 098 Validate Binary Search Tree/validate_binary_search_tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        return helper(root, Int64.MinValue, Int64.MaxValue);
    }
    
    private bool helper(TreeNode root, long minVal, long maxVal) {
        if (root == null) return true;
        if (root.val <= minVal || root.val >= maxVal) return false;
        return helper(root.left, minVal, root.val) && helper(root.right, root.val, maxVal);
    }
} // Author: XC

File: 098 Validate Binary Search Tree/validate_binary_search_tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool dfs(struct TreeNode *node, long curr_min, long curr_max) {
    if (node == NULL) return true;
    if (node->val <= curr_min || node->val >= curr_max) return false;
    return dfs(node->left, curr_min, node->val) && dfs(node->right, node->val, curr_max);
}

bool isValidBST(struct TreeNode* root) {
    return dfs(root, (long)INT_MIN - 1, (long)INT_MAX + 1);
} /* Author: XC */
